src/test/java/com/google/devtools/build/lib/vfs/FileSystemTest.java

```diff
@@ -154,52 +155,57 @@ protected void expectNotFound(Path path) throws IOException {
   }
 
   /**
-   * This function removes an entire directory and all of its contents.
-   * Much like rm -rf directoryToRemove
+   * This function removes an entire directory and all of its contents. Much like rm -rf
+   * directoryToRemove
+   *
+   * <p>This method explicitly only uses Java APIs to interact with files to prevent any issues with
+   * Bazel's own file systems from leaking from one test to another.
    */
-  protected void removeEntireDirectory(File directoryToRemove)
-      throws IOException {
+  protected void removeEntireDirectory(java.nio.file.Path directoryToRemove) throws IOException {
     // make sure that we do not remove anything outside the test directory
     Path testDirPath = testFS.getPath(getTestTmpDir());
-    if (!testFS.getPath(directoryToRemove.getAbsolutePath()).startsWith(testDirPath)) {
+    if (!testFS.getPath(directoryToRemove.toAbsolutePath().toString()).startsWith(testDirPath)) {
       throw new IOException("trying to remove files outside of the testdata directory");
     }
     // Some tests set the directories read-only and/or non-executable, so
     // override that:
-    NativePosixFiles.chmod(
-        directoryToRemove.getPath(),
-        NativePosixFiles.lstat(directoryToRemove.getPath()).getPermissions()
-            | FileStatus.S_IWUSR
-            | FileStatus.S_IXUSR);
-
-    File[] files = directoryToRemove.listFiles();
-    if (files != null) {
-      for (File currentFile : files) {
-        boolean isSymbolicLink = isSymbolicLink(currentFile);
-        if (!isSymbolicLink && currentFile.isDirectory()) {
-          removeEntireDirectory(currentFile);
+    Files.setPosixFilePermissions(
+        directoryToRemove,
+        Sets.union(
+            Files.getPosixFilePermissions(directoryToRemove),
+            ImmutableSet.of(PosixFilePermission.OWNER_WRITE, PosixFilePermission.OWNER_EXECUTE)));
+
+    java.nio.file.Path[] entries = null;
+    try {
+      try (var entriesStream = Files.list(directoryToRemove)) {
+        entries = entriesStream.toArray(java.nio.file.Path[]::new);
+      }
+    } catch (NotDirectoryException ignored) {
+    }
+    if (entries != null) {
+      for (var entry : entries) {
+        boolean isSymbolicLink = Files.isSymbolicLink(entry);
+        if (!isSymbolicLink && Files.isDirectory(entry)) {
+          removeEntireDirectory(entry);
         } else {
           if (!isSymbolicLink) {
-            NativePosixFiles.chmod(
-                currentFile.getPath(),
-                NativePosixFiles.lstat(currentFile.getPath()).getPermissions()
-                    | FileStatus.S_IWUSR);
-          }
-          if (!currentFile.delete()) {
-            throw new IOException("Failed to delete '" + currentFile + "'");
+            Files.setPosixFilePermissions(
+                entry,
+                Sets.union(
+                    Files.getPosixFilePermissions(entry),
+                    ImmutableSet.of(PosixFilePermission.OWNER_WRITE)));
```

Comment:
I don't understand this: on Unix you don't need write permission on a file to delete it, only on the parent directory. (I think it must be a Windows thing?)